### PR TITLE
Update DORT to follow PyTorch changes

### LINKED ISSUE #16355: [Training] DORT fails with new PyTorch changes
### Describe the issue

Per #16353, `Orttraining Linux Lazy Tensor CI Pipeline` fails due to PyTorch's change. Please investigate and fix DORT.

### To reproduce

Build from source and re-run the test.

### Urgency

_No response_

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

main branch

### PyTorch Version

main branch

### Execution Provider

Other / Unknown

### Execution Provider Library Version

_No response_

### PR DESCRIPTION
**Pull request by wschin**
Fix #16355. The root cause change in PyTorch is [#103302](https://github.com/pytorch/pytorch/pull/103302), which seem blocking calling make_fx inside a dynamo backend.

Changes:
1. Move decomposition to `register_backend.py`, so we don't have to call `make_fx` inside DORT, which triggers a bunch of new exceptions.
2. Remove shape inference based on FakeTensorProp since the FX graph received from dynamo contains all shapes now.
3. Fix a macro bug so that DORT can build without CUDA.

Before (3),
```
#if defined(USE_CUDA) || defined(USE_ROCM)
  virtual PhiloxGenerator& PhiloxGenerator__Default() = 0;
#ifdef ENABLE_TRAINING_TORCH_INTEROP
...
#endif
#endif
```
After (3),
```
#if defined(USE_CUDA) || defined(USE_ROCM)
  virtual PhiloxGenerator& PhiloxGenerator__Default() = 0;
#endif
#ifdef ENABLE_TRAINING_TORCH_INTEROP
...
#endif
```
The later one looks better since the `ENABLE_TRAINING_TORCH_INTEROP` is for Python bridge code, not for random-number-generating kernels `PhiloxGenerator`.

### CODE REVIEW

File: onnxruntime/core/session/provider_bridge_ort.cc [hunk 1]
```diff
@@ -1074,8 +1074,8 @@ struct ProviderHostImpl : ProviderHost {
 #endif
 
 #if defined(USE_CUDA) || defined(USE_ROCM)
-
   PhiloxGenerator& PhiloxGenerator__Default() override { return PhiloxGenerator::Default(); }
+#endif
```
Comment (askhade) [inline onnxruntime/core/session/provider_bridge_ort.cc:1078]: Can you explain why this change is needed?

**Comment by askhade** [comment]
Please update the PR description to explain the fix. It will help to review the code as I am not familiar with the code under  torchdynamo dir.

File: onnxruntime/core/session/provider_bridge_ort.cc (hunk 1 above)
Comment (wschin) [reply onnxruntime/core/session/provider_bridge_ort.cc:1078]: Otherwise, DORT won't build for CPU. I think this is an incorrect change introduced by `git merge`, `PhiloxGenerator__Default`, a kernel specific class, shouldn't be related python bridge code such as tracking python object's reference count.

**Comment by wschin** [comment]
> Please update the PR description to explain the fix. It will help to review the code as I am not familiar with the code under torchdynamo dir.

Sure. It's done now.
